src/VisualStudio/Core/Def/Telemetry/VisualStudioWorkspaceTelemetryService.cs

```diff
@@ -7,30 +7,50 @@
 using System.Diagnostics;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
 using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.CodeAnalysis.Internal.Log;
 using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.ProjectTelemetry;
 using Microsoft.CodeAnalysis.Remote;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
 using Microsoft.CodeAnalysis.Telemetry;
 using Microsoft.VisualStudio.Telemetry;
 using Roslyn.Utilities;
 
 namespace Microsoft.VisualStudio.LanguageServices.Telemetry
 {
     [ExportWorkspaceService(typeof(IWorkspaceTelemetryService)), Shared]
-    internal sealed class VisualStudioWorkspaceTelemetryService : AbstractWorkspaceTelemetryService
+    internal sealed partial class VisualStudioWorkspaceTelemetryService : AbstractWorkspaceTelemetryService
     {
         private readonly VisualStudioWorkspace _workspace;
         private readonly IGlobalOptionService _optionsService;
+        private readonly IThreadingContext _threadingContext;
+        private readonly IAsynchronousOperationListener _asynchronousOperationListener;
+
+        /// <summary>
+        /// Queue where we enqueue the information we get from OOP to process in batch in the future.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<ProjectTelemetryData> _workQueue;
 
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public VisualStudioWorkspaceTelemetryService(
             VisualStudioWorkspace workspace,
-            IGlobalOptionService optionsService)
+            IGlobalOptionService optionsService,
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListenerProvider listenerProvider)
         {
             _workspace = workspace;
             _optionsService = optionsService;
+            _threadingContext = threadingContext;
+            _asynchronousOperationListener = listenerProvider.GetListener(FeatureAttribute.Telemetry);
+            InitializeTelemetrySession(TelemetryService.DefaultSession);
```

Comment:
checked with vs perf & rel who own the TelemetryService.DefaultSession bit - the current guidance is that accessing the default session is safe on a background thread, will continue  with that unless I hear otherwise

Reply:
Can you request this information be added here?
https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.telemetry.telemetryservice.defaultsession?view=visualstudiosdk-2019

If it's not in the public documentation, it didn't happen.

Reply:
@dibarbet if I recall we had some surprises where it was safe to use on a background thread _assuming somebody else had read it first_ -- is that still the case?

Reply:
So the previous issues were with MPF and when Roslyn had an older mpf version and initialized it first.  When the original bug was created we were using `TelemetryHelper.DefaultTelemetrySession`.  We're now using `TelemetryService.DefaultSession` which should be safe.  I also tried out the scenarios mentioned in the bug (creating a C++ UWP app and opening the xaml page, then a C# app) and everything appears to be working.

cc also @davkean for Sam's request for updated documentation.

Reply:
We're mixing concepts; one is MPF (the bug you hit), the other is Microsoft.VisualStudio.Telemetry.